File: 02 gpio/MCP23008/mcp23008.c
#include <stdio.h>
#include <wiringPi.h>
#include <mcp23017.h>

int main (void)
{
	int i, bit, ret ;

	// wiringPi 초기화 
	// 이 예제에서 넘버링 방법은 중요하지 않다.
	wiringPiSetupGpio() ;

	// 확장 GPIO 핀 번호를 100번 부터 사용하며 I2C 주소는 0X20임을 알려줌
	// 이 함수를 호출하고 나면 100 ~ 107번까지의 GPIO 핀을 파이의 GPIO 핀과 동일하게 사용할 수 있음
	mcp23008Setup (100, 0x20) ;

	printf ("MCP23008 Test\n") ;

	// 100 ~ 103번까지의 GPIO 핀을 출력 모드로 설정
	for (i = 0 ; i < 4 ; ++i)
	pinMode (100 + i, OUTPUT) ;

	// 104 ~ 107번까지의 GPIO 핀을 입력 모드로 설정
	for (i = 4 ; i < 8 ; ++i){
		pinMode (100 + i, INPUT) ;
	}

	i = 0;
	while(1)
	{
		i++;
		// 100 ~ 103번까지의 GPIO 핀을 ON, OFF 시킴. LED가 켜졌다 꺼짐
		for (bit = 0 ; bit < 4 ; ++bit){
			digitalWrite (100 + bit, (i % 2));
			printf("GPIO:%d Write:%d\n", 100 + bit, (i % 2));
		}
		delay (5);
		// 104 ~ 107번까지의 GPIO 핀을 이용해 100 ~ 103번 핀의 상태를 읽어들임
		for (bit = 4 ; bit < 8 ; ++bit){
			ret = digitalRead (100 + bit);
			printf("GPIO:%d READ:%d\n", 100 + bit, ret);
		}
		delay (3000) ;
	}
	return 0 ;
}

File: 02 gpio/MCP23008/mcp23008.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-
#If this code works, it was written by Seunghyun Lee(www.bluebaynetworks.co.kr).
#If not, I don't know who wrote it

#다운로드 받은 Adafruit_MCP230xx.py 파일에서 Adafruit_MCP230XX를 임포트한다.
from Adafruit_MCP230xx import Adafruit_MCP230XX
from time import sleep


#MCP23008을 초기화 한다.
mcp23008 = Adafruit_MCP230XX(0x20, 8)

for j in range(0, 4):
  mcp23008.config(j, mcp23008.OUTPUT)
for j in range(4, 8):
  mcp23008.config(j, mcp23008.INPUT)

i = 0
try:
  while (True):
    i += 1
    for j in range(0, 4):
      mcp23008.output(j, (i % 2))
      print"GPIO:", j, " WRITE:", (i % 2);

    sleep(0.005)		
    for j in range(4, 8):
      ret = mcp23008.input(j)
      ret = ret >> j
      print"GPIO:", j, " READ:", ret ;

except KeyboardInterrupt:   
	print "MCP23008 Test End" 
	GPIO.cleanup()

sleep(3)		